## Dual-Event Probability Situations

- **Union (Inclusive OR):** At least one of the events occurs.  
- **Exclusive OR (XOR):** Exactly one event occurs, not both.  
- **Intersection (AND):** Both events occur.  
- **Conditional Probability:** Likelihood of one event given the other.  
- **Law of Total Probability:** Overall probability expressed as a sum across all possible cases.  
- **Bayes’ Theorem:** Updates belief about an event after observing evidence.  


#### 1. Event A or Event B:
Union of 2 events \(A\) and \(B\):

**Inclusive or:**
$$
P(A \text{ or } B) = P(A \cup B) = P(A) + P(B) - P(A \cap B)
$$
- We have to subtract out one instance of A and B to avoid double counting
- For example, if we want to calculate the probability of drawing a King or a Heart, the King of Hearts will be included in both the set of Kings and the set of Hearts

In [2]:
p_heart = 13/52 # includes king of hearts
p_king = 4/52 # includes king of hearts
p_heart_and_king = 1/52 # king of hearts
p_heart_or_king = p_heart + p_king - p_heart_and_king # remove 1 king of hearts b/c counted twice
print(p_heart_or_king)

0.3076923076923077


**Exclusive or:**
$$
P(A \text{ xor } B) = P(A \oplus B) = P(A) + P(B) - 2P(A \cap B)
$$
- We have to subtract out all instances of A and B to respect the exclusive or
- For example, if we want to calculate the probability of drawing a King or a Heart, but not both, the King of Hearts should be removed from both the set of Kings and the set of Hearts

In [3]:
p_heart = 13/52 # includes king of hearts
p_king = 4/52 # includes king of hearts
p_heart_and_king = 1/52 # king of hearts
p_heart_or_king = p_heart + p_king - 2 * p_heart_and_king # remove both king of hearts b/c exclusive or
print(p_heart_or_king)

0.28846153846153844


If \(A\) and \(B\) are **mutually exclusive** (cannot occur together):  
$$
P(A \cap B) = 0
$$

$$
P(A \cup B) = P(A) + P(B)
$$

- For example, what's the probability of drawing a Queen or a King
- A card cannot be a Queen and a King, therefore `P(K ∩ Q) = 0`

In [7]:
# What's the probability of drawing a Queen or a King?
p_king = 4/52
p_queen = 4/52
p_king_and_queen = 0/52 # does not exist
p_king_or_queen = p_king + p_queen - p_king_and_queen

print(p_king_or_queen)
print(p_king + p_queen == p_king + p_queen - p_king_and_queen)

0.15384615384615385
True


#### 2. Event A and Event B
Intersection of Two Events (Multiplication Rule)

If \(A\) and \(B\) are **independent**:  

$$
P(A \cap B) = P(A) \cdot P(B)
$$

In [10]:
# Probability of flipping 2 heads on a fair coin
p_head = 1/2
p_2_heads = p_head * p_head
print(p_2_heads)

0.25


If A and B are Dependent
For any two events \(A\) and \(B\):  

$$
P(A \cap B) = P(A) \cdot P(B|A) = P(B) \cdot P(A|B)
$$

In [13]:
# Example: Probability it rains and there is thunder (dependent events)
# Let P(Rain) = 0.3
# Let P(Thunder | Rain) = 0.6

p_rain = 0.3
p_thunder_given_rain = 0.6

# Probability it rains and there is thunder:
p_rain_and_thunder = p_rain * p_thunder_given_rain
print(p_rain_and_thunder)

0.18


In [9]:
# Example: Probability of drawing an Ace, then a King, without replacement from a standard deck
p_ace_first = 4/52  # Probability first card is an Ace
p_king_second_given_ace_first = 4/51  # Probability second card is a King, given first was an Ace

# Since the events are dependent (no replacement), multiply directly:
p_ace_then_king = p_ace_first * p_king_second_given_ace_first

print(p_ace_then_king)

0.006033182503770739


#### 3. Conditional Probability (Review)
Probability of \(A\) given \(B\):  

$$
P(A|B) = \frac{P(A \cap B)}{P(B)}, \quad P(B) > 0
$$

In [14]:
# Example: Conditional probability with cards
# Let A = drawing a heart, B = drawing a king
# What is the probability that a card is a heart given that it is a king?

# We use: P(A|B) = P(A ∩ B) / P(B)
p_king = 4/52
p_heart_and_king = 1/52 # only the king of hearts
p_heart_given_king = p_heart_and_king / p_king
print(f"P(Heart | King) = {p_heart_given_king:.2f}")

P(Heart | King) = 0.25


#### 4. Law of Total Probability
If events B_1, B_2, ..., B_n form a partition of the sample space (mutually exclusive and exhaustive):  

$$
P(A) = \sum_{i=1}^n P(A|B_i) \cdot P(B_i)
$$

In [15]:
# Law of Total Probability with 3 groups

# Probabilities of groups
P_G1 = 0.2   # Students
P_G2 = 0.5   # Workers
P_G3 = 0.3   # Retirees

# Conditional probabilities of purchase given group
P_Purchase_G1 = 0.3
P_Purchase_G2 = 0.6
P_Purchase_G3 = 0.4

# Law of Total Probability
P_Purchase = (
    P_Purchase_G1 * P_G1 +
    P_Purchase_G2 * P_G2 +
    P_Purchase_G3 * P_G3
)

print(f"Overall probability of purchase: {P_Purchase:.2f}")


Overall probability of purchase: 0.48


#### 5. Bayes’ Theorem
For events \(A\) and \(B\) with \(P(B) > 0\):  

$$
P(A|B) = \frac{P(B|A) \cdot P(A)}{P(B)}
$$

With the law of total probability in the denominator:  

$$
P(A|B) = \frac{P(B|A) \cdot P(A)}{\sum_{i=1}^n P(B|A_i) \cdot P(A_i)}
$$

In [16]:
# Bayes' Theorem Example: Customer Groups

# Prior probabilities of groups
P_G1 = 0.2   # Students
P_G2 = 0.5   # Workers
P_G3 = 0.3   # Retirees

# Likelihoods: probability of purchase given group
P_Purchase_G1 = 0.3
P_Purchase_G2 = 0.6
P_Purchase_G3 = 0.4

# Law of Total Probability to find overall P(Purchase)
P_Purchase = (
    P_Purchase_G1 * P_G1 +
    P_Purchase_G2 * P_G2 +
    P_Purchase_G3 * P_G3
)

# Bayes' theorem: posterior probability that customer is a Worker given a purchase
P_G2_given_Purchase = (P_Purchase_G2 * P_G2) / P_Purchase

print(f"Overall probability of purchase: {P_Purchase:.2f}")
print(f"Probability customer is a Worker given a purchase: {P_G2_given_Purchase:.3f}")


Overall probability of purchase: 0.48
Probability customer is a Worker given a purchase: 0.625


### Multi-Event Probability Practice Questions

---

#### 1. Card Draw — Union of Events
**Question:** From a standard 52-card deck, what is the probability of drawing a **heart** or a **king**?

**Answer:**  
$$
P(\text{Heart}) = \tfrac{13}{52}, P(\text{King}) = \tfrac{4}{52}, P(\text{Heart} \cap \text{King}) = \tfrac{1}{52}
$$

$$
P(\text{Heart} \cup \text{King}) = P(\text{Heart}) + P(\text{King}) - P(\text{Heart} \cap \text{King})
$$

$$
= \frac{13}{52} + \frac{4}{52} - \frac{1}{52} = \frac{16}{52} = \frac{4}{13}
$$

---

#### 2. Coin Toss — At Least One Success
**Question:** Two fair coins are flipped. What is the probability of getting **at least one head**?

**Answer:**  
Use complement rule:  

$$
P(\text{At least one head}) = 1 - P(\text{No heads})
$$

$$
= 1 - P(\text{Both tails}) = 1 - \left(\frac{1}{2} \cdot \frac{1}{2}\right) = 1 - \frac{1}{4} = \frac{3}{4}
$$

---

#### 3. Dice Roll — Intersection of Independent Events
**Question:** Roll two fair six-sided dice. What is the probability that the **first die is a 4** and the **second die is even**?

**Answer:**  
Independent events:  

$$
P(\text{First = 4}) = \frac{1}{6}, \quad P(\text{Second even}) = \frac{3}{6} = \frac{1}{2}
$$

$$
P(\text{First = 4} \cap \text{Second even}) = \frac{1}{6} \cdot \frac{1}{2} = \frac{1}{12}
$$

---

#### 4. Law of Total Probability — Email Classification
**Question:** An email system classifies messages as **spam** (30%) or **not spam** (70%).  
- If spam: 90% chance it contains the word "offer".  
- If not spam: 5% chance it contains "offer".  
What is the probability a randomly chosen email contains "offer"?

**Answer:**  

$$
P(\text{Offer}) = P(\text{Offer | Spam}) \cdot P(\text{Spam}) + P(\text{Offer | Not Spam}) \cdot P(\text{Not Spam})
$$

$$
= (0.9)(0.3) + (0.05)(0.7) = 0.27 + 0.035 = 0.305
$$

---

#### 5. Bayes’ Theorem — Inverse Probability
**Question:** Using the email system above, if an email contains "offer", what is the probability it is spam?

**Answer:**  

$$
P(\text{Spam | Offer}) = \frac{P(\text{Offer | Spam}) \cdot P(\text{Spam})}{P(\text{Offer})}
$$

$$
= \frac{0.9 \cdot 0.3}{0.305} = \frac{0.27}{0.305} \approx 0.885
$$

---


### Fixed Example: Conditional Probability with Rain and Thunder
From before, if we know the probability of rain is 0.3, the probability of thunder is 0.2, and the probability there is thunder given that it is raining is 0.6, then what is the probability there is rain given there is thunder?

---

#### Step 1. Define events

Given:
- P(Rain) = 0.3
- P(Thunder | Rain) = 0.6
- P(Thunder) = 0.2
- Want: P(Rain | Thunder)

---

#### Step 2. Compute the intersection
$$
P(\text{Rain} \cap \text{Thunder}) = P(\text{Rain}) \cdot P(\text{Thunder | Rain})
$$

$$
P(\text{Rain} \cap \text{Thunder}) = 0.3 \cdot 0.6 = 0.18
$$

---

#### Step 3. Use the multiplication rule
$$
P(\text{Rain} \cap \text{Thunder}) = P(\text{Thunder}) \cdot P(\text{Rain | Thunder})
$$

$$
0.18 = 0.2 \cdot P(\text{Rain | Thunder})
$$

---

#### Step 4. Solve for \(P(Rain|Thunder)\)
$$
P(\text{Rain | Thunder}) = \frac{0.18}{0.2} = 0.9
$$

---

### ✅ Final Answer
If it rains 30% of days, thunder happens on 20% of days, and thunder given rain is 60%, then:

$$
P(\text{Rain | Thunder}) = 0.9
$$


### Example: Law of Total Probability

#### Problem
On any given day:
- Probability of rain is \(P(A) = 0.3\).  
- If it rains, the probability of thunder is \(P(B|A) = 0.6\).  
- If it does **not** rain, the probability of thunder is \(P(B|A^c) = 0.05\).  

**Question:** What is the overall probability of thunder, \(P(B)\)?  

---

#### Solution
By the Law of Total Probability:

$$
P(B) = P(B|A) \cdot P(A) + P(B|A^c) \cdot P(A^c)
$$

Compute each term:

$$
P(B|A) \cdot P(A) = 0.6 \cdot 0.3 = 0.18
$$

$$
P(B|A^c) \cdot P(A^c) = 0.05 \cdot 0.7 = 0.035
$$

Add them together:

$$
P(B) = 0.18 + 0.035 = 0.215
$$

---

The overall probability of thunder is:
$$
P(B) = 0.215
$$
